pkg/security/admission/admission.go

```diff
@@ -5,83 +5,55 @@ import (
 	"io"
 	"sort"
 	"strings"
+	"sync"
 
+	oscache "github.com/openshift/origin/pkg/client/cache"
+	oadmission "github.com/openshift/origin/pkg/cmd/server/admission"
+	"github.com/openshift/origin/pkg/controller/shared"
+	oscc "github.com/openshift/origin/pkg/security/scc"
 	kadmission "k8s.io/kubernetes/pkg/admission"
 	kapi "k8s.io/kubernetes/pkg/api"
-	"k8s.io/kubernetes/pkg/api/errors"
-	"k8s.io/kubernetes/pkg/client/cache"
 	clientset "k8s.io/kubernetes/pkg/client/clientset_generated/internalclientset"
-	"k8s.io/kubernetes/pkg/runtime"
 	sc "k8s.io/kubernetes/pkg/securitycontext"
 	scc "k8s.io/kubernetes/pkg/securitycontextconstraints"
 	"k8s.io/kubernetes/pkg/serviceaccount"
 	"k8s.io/kubernetes/pkg/util/sets"
-	"k8s.io/kubernetes/pkg/watch"
 
 	allocator "github.com/openshift/origin/pkg/security"
 	"github.com/openshift/origin/pkg/security/uid"
-	"k8s.io/kubernetes/pkg/auth/user"
 	"k8s.io/kubernetes/pkg/util/validation/field"
 
 	"github.com/golang/glog"
 )
 
+var constraintAdmitter *constraint
+var once sync.Once
+
 func init() {
 	kadmission.RegisterPlugin("SecurityContextConstraint", func(client clientset.Interface, config io.Reader) (kadmission.Interface, error) {
-		constraintAdmitter := NewConstraint(client)
-		constraintAdmitter.Run()
+		constraintAdmitter = NewConstraint(client)
 		return constraintAdmitter, nil
 	})
 }
 
 type constraint struct {
 	*kadmission.Handler
-	client clientset.Interface
-
-	reflector *cache.Reflector
-	stopChan  chan struct{}
-	store     cache.Store
+	client    clientset.Interface
+	sccLister *oscache.IndexerToSecurityContextConstraintsLister
 }
 
 var _ kadmission.Interface = &constraint{}
+var _ = oadmission.WantsInformers(&constraint{})
 
 // NewConstraint creates a new SCC constraint admission plugin.
 func NewConstraint(kclient clientset.Interface) *constraint {
-	store := cache.NewStore(cache.MetaNamespaceKeyFunc)
-	reflector := cache.NewReflector(
-		&cache.ListWatch{
-			ListFunc: func(options kapi.ListOptions) (runtime.Object, error) {
-				return kclient.Core().SecurityContextConstraints().List(options)
-			},
-			WatchFunc: func(options kapi.ListOptions) (watch.Interface, error) {
-				return kclient.Core().SecurityContextConstraints().Watch(options)
-			},
-		},
-		&kapi.SecurityContextConstraints{},
-		store,
-		0,
-	)
-
-	return &constraint{
-		Handler: kadmission.NewHandler(kadmission.Create, kadmission.Update),
-		client:  kclient,
-
-		store:     store,
-		reflector: reflector,
-	}
-}
-
-func (a *constraint) Run() {
-	if a.stopChan == nil {
-		a.stopChan = make(chan struct{})
-		a.reflector.RunUntil(a.stopChan)
-	}
-}
-func (a *constraint) Stop() {
-	if a.stopChan != nil {
-		close(a.stopChan)
-		a.stopChan = nil
-	}
+	once.Do(func() {
+		constraintAdmitter = &constraint{
+			Handler: kadmission.NewHandler(kadmission.Create, kadmission.Update),
+			client:  kclient,
+		}
+	})
+	return constraintAdmitter
```

Comment:
> I in test-integration I had some panic since only one of the two constraints had sccLister (set via SetInformer).

Which test-integration test in particular?  I think I'd like to pull the branch look for a different factorization.